internal/api/impl/v1/project/service.go

```diff
@@ -31,49 +37,86 @@ import (
 
 type service struct {
 	project.Service
-	dao           project.DAO
-	folderDAO     folder.DAO
-	datasourceDAO datasource.DAO
-	dashboardDAO  dashboard.DAO
-	secretDAO     secret.DAO
-	variableDAO   variable.DAO
+	dao            project.DAO
+	folderDAO      folder.DAO
+	datasourceDAO  datasource.DAO
+	dashboardDAO   dashboard.DAO
+	roleDAO        role.DAO
+	roleBindingDAO rolebinding.DAO
+	secretDAO      secret.DAO
+	variableDAO    variable.DAO
+	rbac           authorization.RBAC
 }
 
-func NewService(dao project.DAO, folderDAO folder.DAO, datasourceDAO datasource.DAO, dashboardDAO dashboard.DAO, secretDAO secret.DAO, variableDAO variable.DAO) project.Service {
+func NewService(dao project.DAO, folderDAO folder.DAO, datasourceDAO datasource.DAO, dashboardDAO dashboard.DAO, roleDAO role.DAO, roleBindingDAO rolebinding.DAO, secretDAO secret.DAO, variableDAO variable.DAO, rbac authorization.RBAC) project.Service {
 	return &service{
-		dao:           dao,
-		folderDAO:     folderDAO,
-		datasourceDAO: datasourceDAO,
-		dashboardDAO:  dashboardDAO,
-		secretDAO:     secretDAO,
-		variableDAO:   variableDAO,
+		dao:            dao,
+		folderDAO:      folderDAO,
+		datasourceDAO:  datasourceDAO,
+		dashboardDAO:   dashboardDAO,
+		roleDAO:        roleDAO,
+		roleBindingDAO: roleBindingDAO,
+		secretDAO:      secretDAO,
+		variableDAO:    variableDAO,
+		rbac:           rbac,
 	}
 }
 
-func (s *service) Create(entity api.Entity) (interface{}, error) {
+func (s *service) Create(ctx apiInterface.PersesContext, entity api.Entity) (interface{}, error) {
 	if object, ok := entity.(*v1.Project); ok {
-		return s.create(object)
+		return s.create(object, ctx)
 	}
 	return nil, shared.HandleBadRequestError(fmt.Sprintf("wrong entity format, attempting project format, received '%T'", entity))
 }
 
-func (s *service) create(entity *v1.Project) (*v1.Project, error) {
+// Create default roles and role bindings for the project
+func (s *service) createProjectRoleAndRoleBinding(projectName string, ctx apiInterface.PersesContext) error {
+	owner := utils.DefaultOwnerRole(projectName)
```

Comment:
why not returning directly a pointer ?

Reply:
Changed, imo, I prefer to return struct when are sure the func can't return nil 🤔